### ISSUE TITLE
Recreate results from "Neural Program Synthesis with Priority Queue Training"

### ISSUE THREAD

**Issue by xkapastel** [xkapastel/denshi#2]
[Neural Program Synthesis with Priority Queue Training](https://arxiv.org/abs/1801.03526)  introduced a very simple method of "bootstrapping" a random neural network in to solving basic program synthesis problems. I'd like to recreate these results with ABC instead of Brainfuck, learning basic operations on e.g. sums, products, lists and so on.

**Comment by xkapastel** [comment]
My current thinking is that priority queue training can be explained in terms of a new data structure: an "intensional" set. If I'm understanding these terms correctly, a typical set is an "extensional" set where its members are enumerated within your code:

```
var set = new Set();
set.add(0);
set.add(1);
set.add(3);
set.get(); // can only be 1, 2 or 3
```

The only values in this set are 1, 2, and 3. If you want another value, you need to add it yourself. An intensional set is instead defined by a rule: when you add a value to the set, the values you get out will be more like that value.

```
var set = new GenerativeSet();
set.add(0);
set.add(1);
set.add(3);
set.get(); // can be any value "similar" to 1, 2 or 3
```

If `GenerativeSet` is backed by a neural network as in the paper, then at the end of this loop it will be trained in the manner described by PQT:

```
let quota = defaultQuota;
let set = new GenerativeSet();
while (quota > 0) {
  quota--;
  let candidates = set.get(numCandidates);
  candidates.sort(loss);
  let elite = candidates.slice(numElite);
  set.add(elite);
}
```

**Comment by xkapastel** [comment]
Keeping with the "language" terminology, in F#:

   type IGrammar =
      abstract member Add: Word list -> unit
      abstract member Continue: Word list -> Word list
      abstract member Normalize: int -> int

One important change from the previous sketch: you don't just "get" a string out of the set of all strings, you get a string's continuation with `Continue`. The recent stuff about GPT-2 made me realize that prompting a word model with a prefix string is an important part of the API.
